### PR TITLE
Use running statistics to normalize advantages for PPO and AC

### PR DESCRIPTION
**Pull request by emailweixu**
For some tasks, there are rare but importance events which may not appear in one batch. In these situations, the advantage normalization based on one single batch may not correctly normalize the advantage using the desired statistics. This PR changes the normalization based on the EMA statistics of the past batches.

Also add support for normalizing the weighted sum of the advantage instead of normalization for individual reward dimensions.

### CODE REVIEW

File: alf/algorithms/actor_critic_loss.py [hunk 1]
```diff
@@ -28,15 +28,33 @@
...
+    # since the batch may not be large enough to properly normalize within
+    # the batch.
+    batch_norm.eval()
+    return batch_norm(x)
+
+
 @alf.configurable
 class ActorCriticLoss(Loss):
     def __init__(self,
+                 reward_dim=1,
```
Comment (breakds) [inline alf/algorithms/actor_critic_loss.py:49]: Need to update the docstring for `reward_dim`.

File: alf/algorithms/actor_critic_loss.py [hunk 2]
```diff
@@ -67,6 +85,11 @@ def __init__(self,
             normalize_advantages (bool): If True, normalize advantage to zero
                 mean and unit variance within batch for caculating policy
                 gradient. This is commonly used for PPO.
+            normalize_scalar_advantages (bool): If False, the normalization is
```
Comment (breakds) [inline alf/algorithms/actor_critic_loss.py:89]: I think it is not very clear to the reader about the relationship between `normalize_advantages` and `normalize_scalar_advantages`, especially when both of them are `True`. 

In `__init__` it seems that `normalize_scalar_advantages` supersedes `normalize_advantages`, but in `forward` it seems to suggest the other way around.

File: alf/algorithms/actor_critic_loss.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/actor_critic_loss.py:49]: updated.

File: alf/algorithms/actor_critic_loss.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/actor_critic_loss.py:89]: updated dod-string

File: alf/algorithms/actor_critic_loss.py [hunk 3]
```diff
@@ -81,13 +106,23 @@ def __init__(self,
         self._use_gae = use_gae
         self._lambda = td_lambda
         self._use_td_lambda_return = use_td_lambda_return
+        if normalize_scalar_advantages:
```
Comment (hnyu) [inline alf/algorithms/actor_critic_loss.py:109]: For me, I feel it's more natural that `normalize_advantages=True` is a pre-condition for using `normalize_scalar_advantages` which basically determines how we are going to normalize the advantage: either for individual reward dims or for the combined scalar.

File: alf/algorithms/actor_critic_loss.py [hunk 4]
```diff
@@ -28,15 +28,33 @@
                                  ["pg_loss", "td_loss", "neg_entropy"])
 
 
+def normalize(batch_norm, x):
+    batch_norm.train()
+    momentum = batch_norm.momentum
+    if batch_norm.num_batches_tracked * momentum < 1.0:
+        # First the first few batches, we do cumulative moving average
```
Comment (hnyu) [inline alf/algorithms/actor_critic_loss.py]: First -> For

File: alf/algorithms/actor_critic_loss.py (hunk 3 above)
Comment (emailweixu) [reply alf/algorithms/actor_critic_loss.py:109]: I feel it's cumbersome to need to specify both normalize_advantages=True normalize_scalar_advantages=True

File: alf/algorithms/actor_critic_loss.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/actor_critic_loss.py]: fixed

**Review by hnyu — APPROVED** [review]

File: alf/algorithms/actor_critic_loss.py [hunk 5]
```diff
@@ -81,13 +106,23 @@ def __init__(self,
...
+                affine=False,
+                track_running_stats=True)
+            normalize_advantages = False
+        elif normalize_advantages:
+            self._adv_norm = torch.nn.BatchNorm1d(
+                num_features=reward_dim,
+                eps=1e-8,
+                momentum=advantage_norm_momentum,
+                affine=False,
+                track_running_stats=True)
```
Comment (runjerry) [inline alf/algorithms/actor_critic_loss.py:123]: I found that changing this to True could lead to a mismatch between the layer and the ckpt which does not have keys ``running_mean`` and ``running_var``. I wonder if we might want to keep it False and config in conf files instead?

Comment (emailweixu) [reply alf/algorithms/actor_critic_loss.py:123]: You can try the following:
alf.config('Checkpointer.load', strict=False)

Comment (runjerry) [reply alf/algorithms/actor_critic_loss.py:123]: I gave it a try in both the agent and expert conf files, it does not work out-of-the-box, will dig further to figure out why.
